<!doctype html><html><head><title>Python os walk 的坑</title><meta charset=utf-8><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1,maximum-scale=1"><meta name=description content="A blog maintained by an interesting programmer."><meta name=keywords content="Technology,Code,Program,Linux,"><meta name=author content="Xuanwo"><meta property="og:title" content="Python os walk 的坑"><meta property="og:description" content="A blog maintained by an interesting programmer."><meta property="og:type" content="website"><meta property="og:locale" content="en_US"><meta property="og:url" content="https://xuanwo.io/2018/04/01/python-os-walk/"><link rel="shortcut icon" href=/favicon.ico><link rel=stylesheet href=https://xuanwo.io/css/main.min.be107db044369d658c06efdc8f7f5462fda2b704c0a580ab8c227b3dad6ac109.css integrity="sha256-vhB9sEQ2nWWMBu/cj39UYv2itwTApYCrjCJ7Pa1qwQk="><script data-ad-client=ca-pub-8380875817494486 async src=https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js></script><link href=http://gmpg.org/xfn/11 rel=profile><meta name=generator content="Hugo 0.68.3"></head><body><nav class=nav><div class=nav-container><div class=brand><a href=https://xuanwo.io>Xuanwo's Blog</a></div><div class=links><a class=main-nav-link href=/about/>关于我</a>
<a class=main-nav-link href=/posts/>文章</a></div></div></nav><main id=main-content><article class="single container"><header class=single-header><div class=flex><h1>Python os walk 的坑</h1><div class=post-meta><time class=date datetime=2018-04-01T10:07:00.000+00:00 itemprop=datePublished>2018-04-01</time>
/
<span><a href=#disqus_thread class=comment-link>Comments</a></span></div><div class=tag-container><a style=text-decoration:none href=https://xuanwo.io/tags/python><span>Python</span></a></div></div></header><div class=post><blockquote><p>本文是
<a href=https://xuanwo.io/series/learn-from-bug>Learn from BUG</a> 系列的第 1 篇文章, 共 3 篇。</p></blockquote><p>用户反馈使用在 windows 下使用 <a href=https://github.com/yunify/qsctl>qsctl</a> 上传文件的时候会中断并抛出 <code>UnicodeDecodeError</code> 异常，经过一番调查之后发现居然是 <code>os.walk</code> 的坑。</p><h2 id=定位>定位</h2><p>接到用户的反馈之后，首先尝试进行了复现，最后成功的找到了一个能复现该问题的 case:</p><div class=highlight><pre style=background-color:#f0f0f0;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash>&gt;&gt;&gt; <span style=color:#60a0b0;font-style:italic># create a file with bad name and sync with qsctl 1.7.0 on python 2.7.13:</span>
&gt;&gt;&gt; touch <span style=color:#007020;font-weight:700>$(</span><span style=color:#007020>echo</span> -e <span style=color:#4070a0>&#34;\x8b\x8bThis&#34;</span><span style=color:#007020;font-weight:700>)</span>
&gt;&gt;&gt; qsctl sync ./ qs://xxxxxx
</code></pre></div><p>抛出来的异常如下：</p><div class=highlight><pre style=background-color:#f0f0f0;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python>Traceback (most recent call last):
  File <span style=color:#4070a0>&#34;/home/xuanwo/.pyenv/versions/2.7.13/bin/qsctl&#34;</span>, line <span style=color:#40a070>27</span>, <span style=color:#007020;font-weight:700>in</span> <span style=color:#666>&lt;</span>module<span style=color:#666>&gt;</span>
    sys<span style=color:#666>.</span>exit(main())
  File <span style=color:#4070a0>&#34;/home/xuanwo/.pyenv/versions/2.7.13/bin/qsctl&#34;</span>, line <span style=color:#40a070>24</span>, <span style=color:#007020;font-weight:700>in</span> main
    <span style=color:#007020;font-weight:700>return</span> qingstor<span style=color:#666>.</span>qsctl<span style=color:#666>.</span>driver<span style=color:#666>.</span>main()
  File <span style=color:#4070a0>&#34;/home/xuanwo/.pyenv/versions/2.7.13/lib/python2.7/site-packages/qingstor/qsctl/driver.py&#34;</span>, line <span style=color:#40a070>100</span>, <span style=color:#007020;font-weight:700>in</span> main
    command<span style=color:#666>.</span>main(args[<span style=color:#40a070>2</span>:])
  File <span style=color:#4070a0>&#34;/home/xuanwo/.pyenv/versions/2.7.13/lib/python2.7/site-packages/qingstor/qsctl/commands/base.py&#34;</span>, line <span style=color:#40a070>276</span>, <span style=color:#007020;font-weight:700>in</span> main
    <span style=color:#007020;font-weight:700>return</span> cls<span style=color:#666>.</span>send_request()
  File <span style=color:#4070a0>&#34;/home/xuanwo/.pyenv/versions/2.7.13/lib/python2.7/site-packages/qingstor/qsctl/commands/transfer.py&#34;</span>, line <span style=color:#40a070>546</span>, <span style=color:#007020;font-weight:700>in</span> send_request
    cls<span style=color:#666>.</span>upload_files()
  File <span style=color:#4070a0>&#34;/home/xuanwo/.pyenv/versions/2.7.13/lib/python2.7/site-packages/qingstor/qsctl/commands/transfer.py&#34;</span>, line <span style=color:#40a070>165</span>, <span style=color:#007020;font-weight:700>in</span> upload_files
    <span style=color:#007020;font-weight:700>for</span> rt, dirs, files <span style=color:#007020;font-weight:700>in</span> os<span style=color:#666>.</span>walk(source_path):
  File <span style=color:#4070a0>&#34;/home/xuanwo/.pyenv/versions/2.7.13/lib/python2.7/os.py&#34;</span>, line <span style=color:#40a070>286</span>, <span style=color:#007020;font-weight:700>in</span> walk
    <span style=color:#007020;font-weight:700>if</span> isdir(join(top, name)):
  File <span style=color:#4070a0>&#34;/home/xuanwo/.pyenv/versions/2.7.13/lib/python2.7/posixpath.py&#34;</span>, line <span style=color:#40a070>71</span>, <span style=color:#007020;font-weight:700>in</span> join
    path <span style=color:#666>+=</span>  b
<span style=color:#007020>UnicodeDecodeError</span>: <span style=color:#4070a0>&#39;ascii&#39;</span> codec can<span style=color:#4070a0>&#39;t decode byte 0x8b in position 0: ordinal not in range(128)</span>
</code></pre></div><p>之前写过的一篇关于 Python 字符串的<a href=https://xuanwo.io/2017/01/22/encoding-in-python/>文章</a> 曾经分析过类似的问题，Python 2 在进行字符串比较、拼接、替换时，会进行隐式的类型转换。通过查看 <code>posixpath.py</code> 的源码，可以定位到报错的地方：</p><div class=highlight><pre style=background-color:#f0f0f0;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=color:#007020;font-weight:700>def</span> <span style=color:#06287e>join</span>(a, <span style=color:#666>*</span>p):
    <span style=color:#4070a0>&#34;&#34;&#34;Join two or more pathname components, inserting &#39;/&#39; as needed.
</span><span style=color:#4070a0>    If any component is an absolute path, all previous path components
</span><span style=color:#4070a0>    will be discarded.  An empty last part will result in a path that
</span><span style=color:#4070a0>    ends with a separator.&#34;&#34;&#34;</span>
    path <span style=color:#666>=</span> a
    <span style=color:#007020;font-weight:700>for</span> b <span style=color:#007020;font-weight:700>in</span> p:
        <span style=color:#007020;font-weight:700>if</span> b<span style=color:#666>.</span>startswith(<span style=color:#4070a0>&#39;/&#39;</span>):
            path <span style=color:#666>=</span> b
        <span style=color:#007020;font-weight:700>elif</span> path <span style=color:#666>==</span> <span style=color:#4070a0>&#39;&#39;</span> <span style=color:#007020;font-weight:700>or</span> path<span style=color:#666>.</span>endswith(<span style=color:#4070a0>&#39;/&#39;</span>):
            path <span style=color:#666>+=</span>  b <span style=color:#60a0b0;font-style:italic># This is line 71.</span>
        <span style=color:#007020;font-weight:700>else</span>:
            path <span style=color:#666>+=</span> <span style=color:#4070a0>&#39;/&#39;</span> <span style=color:#666>+</span> b
    <span style=color:#007020;font-weight:700>return</span> path
</code></pre></div><p>在 Python 2 下，str 与 unicode 相加，str 会做一次 decode() 转换为 unicode 再相加。也就是说此处报错是因为 path 和 b 的类型不一致导致出现了本不该出现的一次 decode()。顺着代码继续分析，path 和 b 是上层传入的 top 和 name，而这一层的调用是在 os 包的内部进行的。也就是说，os.walk 在处理过程中并没有严格遵循保持类型一致的不成文约定，而是在传入 unicode 的情况下，出现了 str 类型。知道了问题出在 <code>os.walk</code>，接下来再看看 <code>os.walk</code> 的实现就能明白问题的所在了：</p><div class=highlight><pre style=background-color:#f0f0f0;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-python data-lang=python><span style=color:#007020;font-weight:700>def</span> <span style=color:#06287e>walk</span>(top, func, arg):
  islink, join, isdir <span style=color:#666>=</span> path<span style=color:#666>.</span>islink, path<span style=color:#666>.</span>join, path<span style=color:#666>.</span>isdir

  <span style=color:#60a0b0;font-style:italic># We may not have read permission for top, in which case we can&#39;t</span>
  <span style=color:#60a0b0;font-style:italic># get a list of the files the directory contains.  os.path.walk</span>
  <span style=color:#60a0b0;font-style:italic># always suppressed the exception then, rather than blow up for a</span>
  <span style=color:#60a0b0;font-style:italic># minor reason when (say) a thousand readable directories are still</span>
  <span style=color:#60a0b0;font-style:italic># left to visit.  That logic is copied here.</span>
  <span style=color:#007020;font-weight:700>try</span>:
      <span style=color:#60a0b0;font-style:italic># Note that listdir and error are globals in this module due</span>
      <span style=color:#60a0b0;font-style:italic># to earlier import-*.</span>
      names <span style=color:#666>=</span> listdir(top)
  <span style=color:#007020;font-weight:700>except</span> error, err:
      <span style=color:#007020;font-weight:700>if</span> onerror <span style=color:#007020;font-weight:700>is</span> <span style=color:#007020;font-weight:700>not</span> None:
          onerror(err)
      <span style=color:#007020;font-weight:700>return</span>

  dirs, nondirs <span style=color:#666>=</span> [], []
  <span style=color:#007020;font-weight:700>for</span> name <span style=color:#007020;font-weight:700>in</span> names:
      <span style=color:#007020;font-weight:700>if</span> isdir(join(top, name)):
          dirs<span style=color:#666>.</span>append(name)
      <span style=color:#007020;font-weight:700>else</span>:
          nondirs<span style=color:#666>.</span>append(name)

  <span style=color:#007020;font-weight:700>if</span> topdown:
      <span style=color:#007020;font-weight:700>yield</span> top, dirs, nondirs
  <span style=color:#007020;font-weight:700>for</span> name <span style=color:#007020;font-weight:700>in</span> dirs:
      new_path <span style=color:#666>=</span> join(top, name)
      <span style=color:#007020;font-weight:700>if</span> followlinks <span style=color:#007020;font-weight:700>or</span> <span style=color:#007020;font-weight:700>not</span> islink(new_path):
          <span style=color:#007020;font-weight:700>for</span> x <span style=color:#007020;font-weight:700>in</span> walk(new_path, topdown, onerror, followlinks):
              <span style=color:#007020;font-weight:700>yield</span> x
  <span style=color:#007020;font-weight:700>if</span> <span style=color:#007020;font-weight:700>not</span> topdown:
      <span style=color:#007020;font-weight:700>yield</span> top, dirs, nondirs
</code></pre></div><p>等到读完 <code>os.walk</code> 的实现我们就能明白，<code>os.walk</code> 也是被迫背锅的，那个奇怪的 str 是由 <code>os.listdir</code> 返回的。但是 <code>os.listdir</code> 的实现是系统相关的，<code>os.walk</code> 理应屏蔽掉编码的细节，为用户提供一个行为一致的接口。</p><h2 id=修复>修复</h2><p>定位到问题之后，修复起来就变得简单了。检查一下 listdir 的返回值，如果类型是 str，我们就试着去做一次 decode。如果报错了的话，我们需要通过 onerror 来处理这个情况然后把这个文件从结果中去掉以保证同步可以继续进行。
最后通过这两个 commit 对这个问题进行了修复：</p><ul><li><a href=https://github.com/yunify/qsctl/commit/f071667b12f8172451a9e7d63dcdd44f9348bf22>Handle UnicodeDecodeError while use os.walk</a></li><li><a href=https://github.com/yunify/qsctl/commit/840a97ef8954fbe35659cfc6d457f461dcf2b77d>Handle illegal characters in a better way</a></li></ul><h2 id=总结>总结</h2><ul><li>接口是开发者与用户的神圣契约，我们要尽量避免不一致的行为。</li><li>上层接口要尽可能屏蔽下层的细节，不要把本该自己处理的问题扔给用户处理。</li><li>Python 2 快点死掉吧 = =</li></ul><h2 id=动态>动态</h2><ul><li>正如开头所说的，这篇文章是 “Learn From BUG” 系列的第一篇，之后我会不定期的整理和分享一些自己平时遇到的 BUG 解决思路。一方面是为了自己能够从 BUG 中学到更多，另一方面是希望能够帮助到被类似 BUG 困扰的人们。</li><li>随着再一次的心血来潮，我在 Linode 买了台机器部署并且把 Blog 迁移了过去，去掉了减速 CDN，实测速度比之前快上了不少。</li><li>《比宇宙更遥远的地方》完结撒花了，我现在唯一的感触就是这次的圣地巡礼可能有点贵 = =</li></ul></div><div id=disqus_thread></div><script type=application/javascript async>let disqus_config=function(){this.page.identifier='\/2018\/04\/01\/python-os-walk\/';this.page.title='Python os walk 的坑';this.page.url='https:\/\/xuanwo.io\/2018\/04\/01\/python-os-walk\/';};(function(){let d=document,s=d.createElement('script');s.async=true;s.src='//only0god.disqus.com/embed.js';s.setAttribute('data-timestamp',+new Date());(d.head||d.body).appendChild(s);})();</script></article></main><footer class="footer container"><a class=main-nav-link href=/categories/>分类</a>
<a class=main-nav-link href=/tags/>标签</a>
<a class=main-nav-link href=/series/>专题</a>
<a class=main-nav-link href=/blogroll/>友链</a>
<a class=main-nav-link href=/index.xml>订阅</a>
<script type=application/javascript>var doNotTrack=false;if(!doNotTrack){window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;ga('create','UA-51515330-1','auto');ga('send','pageview');}</script><script async src=https://www.google-analytics.com/analytics.js></script></footer></body></html>